homeassistant/components/sensor/__init__.py

```diff
@@ -476,6 +481,25 @@ def state(self) -> Any:
                     f"Sensor {self.entity_id} provides state value '{value}', "
                     "which is not in the list of options provided"
                 )
+            return value
+
+        # If the sensor has neither a device class, a state class nor
+        # a unit_of measurement then there are no further checks or conversions
+        if not device_class and not state_class and not unit_of_measurement:
+            return value
+
+        if not isinstance(value, (int, float, Decimal)):
+            if value == "":
+                return None
```

Comment:
I added this check mostly for integrations that do string parsing (mqtt, template, rest, scrape...)
I think that when a number is expected, it is safe to assume an empty string is simply the absence of a value.

Reply:
I know of some cases where an empty string is accepted as a valid value (see mqtt string platform). A `None` value represents `STATE_UNKNOWN` which is something different.

Reply:
My adjustment only applies when a valid number is expected (ie. it has a state class, or a uom, or a number device class).

Blank string will still be allowed for other sensors.

Reply:
Why should we convert it to `None` silently? I'd expect us to raise an exception so the integration can be fixed.

Reply:
I've removed this. I had to adjust the "rest" tests accordingly.

Reply:
Had to adjust template also.